### ISSUE TITLE
Latest version throwing com.artipie.asto.factory.StorageNotFoundException: Storage with type 's3' is not found

### ISSUE THREAD

**Issue by prakashdivyy** [artipie/artipie#1244]
Hello team, I'm trying to setup artipie with S3 (minio) on local using artipie latest version. I found that S3 storage is not found.
When I tried with older version (example: v0.22), itcan successfully store the artifact to S3.

```
artipie  | [vert.x-eventloop-thread-0] INFO com.artipie.http.MainSlice - Failure: java.util.concurrent.CompletionException: com.google.common.util.concurrent.UncheckedExecutionException: com.artipie.asto.factory.StorageNotFoundException: Storage with type 's3' is not found.
artipie  |      at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:687)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:144)
artipie  |      at com.artipie.SliceFromConfig.<init>(SliceFromConfig.java:83)
artipie  |      at com.artipie.http.ArtipieRepositories.lambda$resolve$1(ArtipieRepositories.java:111)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:144)
artipie  |      at com.artipie.http.ArtipieRepositories.resolve(ArtipieRepositories.java:103)
artipie  |      at com.artipie.http.ArtipieRepositories.lambda$slice$0(ArtipieRepositories.java:81)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
artipie  |      at com.artipie.http.ArtipieRepositories.slice(ArtipieRepositories.java:77)
artipie  |      at com.artipie.http.SliceByPath.response(SliceByPath.java:85)
artipie  |      at com.artipie.http.DockerRoutingSlice.response(DockerRoutingSlice.java:88)
artipie  |      at com.artipie.http.rt.RtRulePath.response(RtRulePath.java:53)
artipie  |      at com.artipie.http.rt.SliceRoute.lambda$response$0(SliceRoute.java:69)
artipie  |      at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
artipie  |      at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:1002)
artipie  |      at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
artipie  |      at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
artipie  |      at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
artipie  |      at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
artipie  |      at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
artipie  |      at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
artipie  |      at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
artipie  |      at com.artipie.http.rt.SliceRoute.response(SliceRoute.java:72)
artipie  |      at com.artipie.http.Slice$Wrap.response(Slice.java:62)
artipie  |      at com.artipie.http.slice.LoggingSlice.lambda$response$1(LoggingSlice.java:71)
artipie  |      at com.artipie.http.SafeSlice$RsSafe.send(SafeSlice.java:79)
artipie  |      at com.artipie.micrometer.MicrometerSlice$MicrometerResponse.send(MicrometerSlice.java:172)
artipie  |      at com.artipie.vertx.VertxSliceServer.serve(VertxSliceServer.java:190)
artipie  |      at com.artipie.vertx.VertxSliceServer.lambda$proxyHandler$1(VertxSliceServer.java:165)
artipie  |      at io.vertx.lang.rx.DelegatingHandler.handle(DelegatingHandler.java:20)
artipie  |      at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:67)
artipie  |      at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:30)
artipie  |      at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:55)
artipie  |      at io.vertx.core.impl.DuplicatedContext.emit(DuplicatedContext.java:158)
artipie  |      at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:145)
artipie  |      at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:157)
artipie  |      at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
artipie  |      at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.vertx.core.http.impl.Http1xUpgradeToH2CHandler.channelRead(Http1xUpgradeToH2CHandler.java:116)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:336)
artipie  |      at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
artipie  |      at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:444)
artipie  |      at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:280)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.vertx.core.http.impl.Http1xOrH2CHandler.end(Http1xOrH2CHandler.java:61)
artipie  |      at io.vertx.core.http.impl.Http1xOrH2CHandler.channelRead(Http1xOrH2CHandler.java:38)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
artipie  |      at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
artipie  |      at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
artipie  |      at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
artipie  |      at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
artipie  |      at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
artipie  |      at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
artipie  |      at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
artipie  |      at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
artipie  |      at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
artipie  |      at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
artipie  |      at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
artipie  |      at java.base/java.lang.Thread.run(Thread.java:833)
artipie  | Caused by: com.google.common.util.concurrent.UncheckedExecutionException: com.artipie.asto.factory.StorageNotFoundException: Storage with type 's3' is not found.
artipie  |      at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
artipie  |      at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
artipie  |      at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4848)
artipie  |      at com.artipie.settings.cache.CachedStorages.storage(CachedStorages.java:72)
artipie  |      at java.base/java.util.Optional.map(Optional.java:260)
artipie  |      at com.artipie.settings.StorageByAlias.lambda$storage$0(StorageByAlias.java:39)
artipie  |      at java.base/java.util.Optional.map(Optional.java:260)
artipie  |      at com.artipie.settings.StorageByAlias.storage(StorageByAlias.java:38)
artipie  |      at com.artipie.settings.cache.StoragesCache.storage(StoragesCache.java:47)
artipie  |      at com.artipie.settings.repo.RepoConfig.lambda$storageOpt$1(RepoConfig.java:182)
artipie  |      at java.base/java.util.Optional.map(Optional.java:260)
artipie  |      at com.artipie.settings.repo.RepoConfig.storageOpt(RepoConfig.java:177)
artipie  |      at com.artipie.settings.repo.RepoConfig.storage(RepoConfig.java:164)
artipie  |      at com.artipie.SliceFromConfig.build(SliceFromConfig.java:230)
artipie  |      at com.artipie.SliceFromConfig.lambda$new$0(SliceFromConfig.java:84)
artipie  |      at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684)
artipie  |      ... 80 more
artipie  | Caused by: com.artipie.asto.factory.StorageNotFoundException: Storage with type 's3' is not found.
artipie  |      at com.artipie.asto.factory.Storages.newStorage(Storages.java:70)
artipie  |      at com.artipie.settings.cache.CachedStorages.lambda$storage$0(CachedStorages.java:87)
artipie  |      at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4853)
artipie  |      at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
artipie  |      at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
artipie  |      at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
artipie  |      at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
artipie  |      ... 95 more
```

Kindly help,
Regards

**Comment by olenagerasimova** [comment]
hi @prakashdivyy we provided an opportunity to add [custom storage](https://github.com/artipie/artipie/wiki/Configuration-Storage#custom-storage) implementations and separated `s3` storage implementation into separate module as well. You need to download jar with [`s3` implementation from maven](https://search.maven.org/artifact/com.artipie/asto-s3/v1.15.4/jar) and add it to the classpath.

**Comment by prakashdivyy** [comment]
Hi @olenagerasimova , thanks for your answer. Do you have any examples on how to include that class path into docker application?

**Comment by andpopov** [comment]
Hi @prakashdivyy
We prepared Artipie Dockerfile that includes all types of storages(+s3 storage also).
The link is [here](https://github.com/andpopov/artipie/blob/1244_Dockerfile_with_asto_deps/docker-with-deps/Dockerfile)
We would like to ask you to test s3 support.

Steps to build new docker image:
1. Clone the git repository:
``git clone https://github.com/andpopov/artipie.git``
2. Go to repository folder: ``artipie``
3. Switch to branch ``1244_Dockerfile_with_asto_deps``:
``git checkout 1244_Dockerfile_with_asto_deps``
4. Go to folder: ``docker-with-deps``
5. Build new docker image:
docker build -t artipie_s3 .

**Comment by andpopov** [comment]
Hello @prakashdivyy 
Released Artipie [v0.28.2](https://github.com/artipie/artipie/releases/tag/v0.28.2) that contains all storage types (including s3), please pull corresponding docker [image](https://hub.docker.com/layers/artipie/artipie/v0.28.2/images/sha256-77bbcfbfe79c16152ac0eddc71658e0c352169084f5fcf54dece10c67814c883?context=explore)
